src/thor/trace_route_action.cc

```diff
@@ -392,52 +398,53 @@ thor_worker_t::map_match(Api& request, uint32_t best_paths) {
                      m_temp_route_discontinuities);
-    } // trace_route can return multiple trip paths and cannot have discontinuities
+    } // trace_route can return multiple trip paths
     else {
-      auto& route = *request.mutable_trip()->mutable_routes()->Add();
-      auto origin_iter = match_results.begin();
-
-      // the following logic builds legs on edge candidates that generated by meili
-
-      // To build legs, the first valid matched location should be on the first edge candidate
-      while (origin_iter->edgeid != m_temp_path_edges.front().edgeid) {
-        ++origin_iter;
-      }
 
+      auto origin_iter = match_results.begin();
+      /*the following logic builds leges on edge candidates that generated by meili*/
       // The following logic put break points (matches results) on edge candidates to form legs
       // logic assumes the both match results and edge candidates are topologically sorted in correct
       // order
-      int last_leg_index = 0;
-      for (int i = 0, n = static_cast<int>(m_temp_path_edges.size()); i < n; ++i) {
-        const auto& path_edge = m_temp_path_edges[i];
-
-        for (auto iter = origin_iter + 1; iter != match_results.end(); ++iter) {
-          if (path_edge.edgeid != iter->edgeid ||
-              options.shape(iter - match_results.begin()).type() != valhalla::Location::kBreak) {
-            continue;
-          }
 
-          // TODO: if this were a time dependent match and you werent use_timestamps
-          // you need to propogate the date information to each legs origin location
-          // because triplegbuilder relies on it. form path set the first one but
-          // on the subsequent legs we will need to set them by doing time offsetting
-          // like is done in route_action.cc thor_worker_t::depart_at
-
-          Location* origin_location = options.mutable_shape(origin_iter - match_results.begin());
-          Location* destination_location = options.mutable_shape(iter - match_results.begin());
-          add_path_edge(origin_location, *origin_iter);
-          add_path_edge(destination_location, *iter);
-
-          TripLegBuilder::Build(controller, matcher->graphreader(), mode_costing,
-                                m_temp_path_edges.begin() + last_leg_index,
-                                m_temp_path_edges.begin() + i + 1, *origin_location,
-                                *destination_location, std::list<valhalla::Location>{},
-                                *route.mutable_legs()->Add(), interrupt,
-                                &m_temp_route_discontinuities);
-          // beginning of next leg will be the end of this leg
-          origin_iter = iter;
-          // store the starting index of the path_edges
-          last_leg_index = i;
+      for (const vector<PathInfo>& path_infos : m_temp_disjoint_edge_groups) {
+        // for each disjoint edge group, we make a new route for it.
+        // We use multi-route to handle discontinuity
```

Comment:
this is unfortunate because it removes our ability to use the next most probable match as an alternate match in the output. at the moment we only allow multiple probable matches in trace_attributes. there are tests for this but they arent very deep. we should check what happens when there is a discontinuity but also someone has asked for top-k > 1. there are tests in the map matching unit test that test this without discontinuity.